src/rewrite_gelu.cpp

```diff
@@ -33,74 +33,75 @@ namespace migraphx {
+    auto c              = m.insert_instruction(ins, make_op("mul"), b, x);
+    auto u              = m.insert_instruction(ins, make_op("add"), c, xb);
+    auto neg_u          = m.insert_instruction(ins, make_op("neg"), u);
+    auto d              = m.insert_instruction(ins, make_op("sub"), neg_u, u);
+    auto emu            = m.insert_instruction(ins, make_op("exp"), d);
+    auto e              = insert_common_op(m, ins, make_op("add"), {one, emu});
+    auto cdf            = insert_common_op(m, ins, make_op("div"), {one, e});
+    auto y              = m.insert_instruction(ins, make_op("mul"), x, cdf);
+    m.replace_instruction(ins, y);
+}
+
+/**
+ * Finds erfGELU blocks using the Gaussian distribution and replaces them with the tanh_exp
+ * approximation if the data type is fp16. TODO consider also for fp8 datatype.
  */
 struct find_gelu_erf
 {
-    auto matcher() const { return match::gelu_erf(); }
+    auto matcher() const { return match::any_of(match::gelu_erf(), match::gelu_tanh()); }
 
     void apply(module& m, const match::matcher_result& r) const
     {
-        auto ins = r.result;
         auto x   = r.instructions["x"];
-        if(x->get_shape().type() != migraphx::shape::half_type)
+        auto input_type                              = x->get_shape().type();
+        std::set<decltype(input_type)> convert_types = {migraphx::shape::half_type};
+        if(not contains(convert_types, input_type))
             return;
 
-        auto lit = m.add_literal(literal{shape{x->get_shape().type()}, {1.702f}});
-        auto mul = insert_common_op(m, ins, make_op("mul"), {x, lit});
-        auto sig = m.insert_instruction(ins, make_op("neg"), mul);
-        sig      = m.insert_instruction(ins, make_op("exp"), sig);
-        auto one = m.add_literal(literal{shape{x->get_shape().type()}, {1.0f}});
-        sig      = insert_common_op(m, ins, make_op("add"), {sig, one});
-        sig      = m.insert_instruction(ins, make_op("div"), x, sig);
-        m.replace_instruction(ins, sig);
+        replace_with_tanh_exp_gelu(m, r);
     }
 };
 
 /**
- * Find fastGELU blocks (where the graph already does a GELU approximation) and replace them
- * with an alternative approximation that is less likely to overflow.
- * The replacement approximation is equivalent to:
- * GELU(x) ~= 0.5 * x * ( 1 + tanh( sqrt(2/M_PI) * (x + 0.044715 * x^3)))
- * You can rearrange to the form used in this by recognizing that
- * 1 + tanh(x) = (2) / (1 + exp(-2 * x)).
- * The fitting constant 0.044715 is from
- * A. Choudhury, ‘A simple approximation to the area under standard normal curve’, Mathematics and
- * Statistics, vol. 2, no. 3, pp. 147–149, 2014.
+ * Find tanhGELU blocks and replace them with a rearranged version that is less likely to overflow
+ * and is more performant.
  */
 struct find_tanh_fast_gelu
 {
     auto matcher() const { return match::gelu_tanh(); }
 
     void apply(module& m, const match::matcher_result& r) const
```

Comment:
This struct and similar ones with a `matcher()` and `apply()` method seem to be set up for a polymorphic code design, without actually using polymorphism or any other kind of class interrelationship.  Is there a reason it's done this way?

Reply:
> method seem to be set up for a polymorphic code design, without actually using polymorphism

It uses polymorphism. 

> or any other kind of class interrelationship.

You mean a type-erased interface?

>  Is there a reason it's done this way?

Versus type-erasure? Because the matchers can be inlined and with CSE the matchers that check for the same things can be done only once.